### DISCUSSION TITLE
ACK <commit hash> ?

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#963] [Q&A] [answered]
Can someone please explain to me why reviewers add `ACK <commit hash>` on this repo? Reviewers also have to click 'approved' via the GitHub UI. GitHub already removes reviews if the branch is force pushed so adding the explicit commit hash is redundant?

**Comment by apoelstra** [comment, marked as answer]
There are a couple of reasons:

* Our [merge script](https://github.com/bitcoin-core/bitcoin-maintainer-tools/blob/main/github-merge.py) will only acknowledge ACKs on the correct commit
* I sometimes review the wrong commit (e.g. if there was a rebase that I didn't fetch); if I did an unqualified ACK nobody would be able to tell
* ...and in fact, even if I add the wrong commt ID, Github will just accept it. Github is not trustworthy. Its ability to automatically dismiss reviews on push is also intermittent and sketchy.
* If I mark the commit ID that I most recently reviewed, and then things change, it's easy for me to use range-diff to tell what changed. (Though of course Github makes this pointlessly hard by turning commit IDs into hard-to-select links, and also by hiding comments..)

**Reply by Kixunil** [reply]
> Github is not trustworthy.

If you assume this, we should GPG-sign our ACKs. And perhaps stop using sha1. (use `cargo-crev` instead?)

**Reply by apoelstra** [reply]
i like the idea of gpg-signing acks. It's probably a bit too much work and a bit too much noise for me to advise, when it would be very visible if github were to invent ACKs from nowhere (aside from duplicating them sometimes, which it does routinely if you use the website over a bad connection).

On the other hand we *do* routinely see Github hide comments, show commits in the wrong order, make unpredictable and inconsistent decisions about which "acks" count toward the ACK requirement, and most importantly, we see it continue to count ACKs even after the underlying PR has completely changed. While these well-known issues are enough for me to dismiss github as "untrustworthy", I do think they'd be crossing a new line if they were to undermine "`ACK <commit ID>`" somehow.

As for SHA-1, git actually uses [SHA-1DC](https://lwn.net/Articles/898522/) which somehow protects against known collision attacks.

**Reply by apoelstra** [reply]
+1 to "use cargo-crev" but I'm curious what you mean by using it *instead* ... does it support a workflow where you could ack/nack unmerged PRs?

**Reply by Kixunil** [reply]
I meant using it to compute the hash and use that one instead.

**Comment by elichai** [comment]
Also this: https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review

**Comment by sanket1729** [comment]
Also, lets other maintainers make decisions like Person A ACKed the code that only has two lines simple diff with the latest tip. We can assume that A would not complain if we merge this.

**Comment by tcharding** [comment]
> Also, lets other maintainers make decisions like Person A ACKed the code that only has two lines simple diff with the latest tip. We can assume that A would not complain if we merge this.

Oh, cool. I'm glad of this.

Thanks all.

**Comment by sanket1729** [comment]
@tcharding  

An example of Github failing to remove approval after a force push to a different commit. https://github.com/ElementsProject/rust-elements/pull/120.

**Reply by tcharding** [reply]
That is a pretty bad bug.
